crates/pink/sidevm/sidevm/src/net.rs

```diff
@@ -0,0 +1,162 @@
+    type Output = Result<TcpStream>;
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let rv = match ocall::tcp_accept(self.listener.res_id.0) {
+            Ok(rv) => rv,
+            Err(err) => return Poll::Ready(Err(err)),
+        };
+        match rv {
+            env::Poll::Ready(res_id) => Poll::Ready(Ok(TcpStream {
+                res_id: ResourceId(res_id),
+                buf: Default::default(),
+                start: 0,
+                filled: 0,
+            })),
+            env::Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl TcpListener {
+    /// Listen on the specified address for incoming TCP connections.
+    pub async fn listen(addr: &str) -> Result<Self> {
+        let res_id = ResourceId(ocall::tcp_listen(addr.into(), 10)?);
+        Ok(Self { res_id })
+    }
+
+    /// Accept a new incoming connection.
+    pub async fn accept(&self) -> Result<TcpStream> {
+        Acceptor { listener: self }.await
+    }
+}
+
+impl Accept for TcpListener {
+    type Conn = TcpStream;
+
+    type Error = env::OcallError;
+
+    fn poll_accept(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Conn, Self::Error>>> {
+        let mut accept = Acceptor {
+            listener: self.get_mut(),
+        };
+        let x = Pin::new(&mut accept).poll(cx).map(Some);
+        log::info!("Poll accept = {:?}", x);
+        x
+    }
+}
+
+impl AsyncRead for TcpStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> Poll<std::io::Result<()>> {
+        let result = {
+            let size = buf.remaining().min(512);
+            let buf = buf.initialize_unfilled_to(size);
+            ocall::poll_read(self.res_id.0, buf)
+        };
+        match result {
+            Ok(env::Poll::Pending) => Poll::Pending,
+            Ok(env::Poll::Ready(len)) => {
+                let len = len as usize;
+                if len > buf.remaining() {
+                    Poll::Ready(Err(Error::from_raw_os_error(
+                        env::OcallError::InvalidEncoding as i32,
+                    )))
+                } else {
+                    buf.advance(len);
+                    Poll::Ready(Ok(()))
+                }
+            }
+            Err(err) => Poll::Ready(Err(Error::from_raw_os_error(err as i32))),
+        }
+    }
+}
+
+impl AsyncBufRead for TcpStream {
+    fn poll_fill_buf(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<std::io::Result<&[u8]>> {
+        if self.filled == self.start {
+            match ocall::poll_read(self.res_id.0, &mut self.buf[..]) {
+                Ok(env::Poll::Pending) => return Poll::Pending,
+                Err(err) => return Poll::Ready(Err(Error::from_raw_os_error(err as i32))),
+                Ok(env::Poll::Ready(sz)) => {
+                    self.filled = sz as _;
```

Comment:
filled is always the last read bytes into the buffer.

Reply:
Got it. So here since `self.filled == self.start`, we can read to the head of the buffer

Reply:
In fact, only when both of them are `0` they can eq to each other. Because when `start` increase to `filled`, both are reset to `0`:
```rust
    fn consume(mut self: Pin<&mut Self>, amt: usize) {
        self.start += amt;
        if self.start >= self.filled {
            self.start = 0;
            self.filled = 0;
        }
    }
```

Maybe I should change the condition to `if self.filled == 0` to clear.

Reply:
I would write `if self.filled == 0 && self.start == 0` and add a comment that once filled == start it will be always set to 0

Reply:
Good to comment it.